### PR TITLE
feat(api): create save challenge route

### LINKED ISSUE #49932: POST /challenge/save-challenge

### PR DESCRIPTION
**Pull request by Sboonny**
Checklist:

- [ ] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [ ] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [ ] My pull request targets the `main` branch of freeCodeCamp.
- [ ] I have tested these changes either locally on my machine, or GitPod.

Closes #49932

### CODE REVIEW

File: api/src/routes/challenge.test.ts [hunk 1]
```diff
@@ -538,6 +553,46 @@ describe('challengeRoutes', () => {
...
+          setCookies
+        }).send({ savedChallenges: stepChallenge.id });
+
+        expect(response.statusCode).toEqual(400);
+        expect(response.body).toEqual({
+          code: 'FST_ERR_VALIDATION',
+          error: 'Bad Request',
+          message: "body must have required property 'id'",
+          statusCode: 400
+        });
```
Comment (raisedadead) [inline api/src/routes/challenge.test.ts:592]: Make this consistent with the current API

**Review by raisedadead — CHANGES_REQUESTED** [review]

File: api/src/routes/challenge.ts [hunk 2]
```diff
@@ -379,5 +379,48 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
...
+        }
+      }
+    },
+    async (req, reply) => {
+      try {
+        const saveChallenge = await fastify.prisma.user.update({
+          where: { id: req.session.user.id },
+          select: {
+            savedChallenges: true
+          },
```
Comment (Sboonny) [inline api/src/routes/challenge.ts:404]: this isn't query the challenges, this is query the user if they have savedchallenge, use utili.

File: api/src/routes/challenge.ts [hunk 3]
```diff
@@ -403,26 +402,35 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
         const user = await fastify.prisma.user.findUniqueOrThrow({
           where: { id: req.session.user.id }
         });
+        const savableChallenges = user.savedChallenges;
```
Comment (ojeytonwilliams) [inline api/src/routes/challenge.ts:405]: This is the currently saved challenges, not the challenges that can be saved. Those are challenges with type `multifileCertProject`

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: api/src/routes/challenge.ts (hunk 3 above)
Comment (Sboonny) [reply api/src/routes/challenge.ts:411]: yeab https://github.com/Sboonny/freeCodeCamp/blob/a19535171796fa8103ceb640fe86d3321419c544/api-server/src/server/boot/challenge.js#L127-L129, I am trying to figure out why the body isn't updated after the endpoint has run.

File: api/src/utils/common-challenge-functions.ts [hunk 4]
```diff
@@ -65,6 +61,41 @@ export type CompletedChallenge = {
...
+ * submission endpoints.
+ *
+ * @param challengeId The id of the submitted challenge.
+ * @param user The existing user record.
+ * @param challenge The saveble challenge.
+ * @returns Saved challenges.
+ */
+export function saveUserChallengeData(
+  challengeId: string,
+  user: user,
```
Comment (ojeytonwilliams) [inline api/src/utils/common-challenge-functions.ts:75]:
```suggestion
  savedChallenges: SavedChallenge[],
```

File: api/src/utils/common-challenge-functions.ts [hunk 5]
```diff
@@ -65,6 +61,41 @@ export type CompletedChallenge = {
   files?: CompletedChallengeFile[];
 };
 
+/**
+ * Helper function to save a user's challenge data. Used in challenge
+ * submission endpoints.
+ *
+ * @param challengeId The id of the submitted challenge.
+ * @param user The existing user record.
```
Comment (ojeytonwilliams) [inline api/src/utils/common-challenge-functions.ts:69]:
```suggestion
 * @param savedChallenges The user's saved challenges.
```

File: api/src/utils/common-challenge-functions.ts [hunk 6]
```diff
@@ -65,6 +61,41 @@ export type CompletedChallenge = {
...
+) {
+  const challengeToSave: SavedChallenge = {
+    id: challengeId,
+    lastSavedDate: Date.now(),
+    files: challenge.files?.map(file =>
+      pick(file, ['contents', 'key', 'name', 'ext', 'history'])
+    ) as SavedChallengeFile[]
+  };
+
+  const savedIndex = user.savedChallenges.findIndex(
```
Comment (ojeytonwilliams) [inline api/src/utils/common-challenge-functions.ts:86]:
```suggestion
  const savedIndex = savedChallenges.findIndex(
```

File: api/src/utils/common-challenge-functions.ts [hunk 7]
```diff
@@ -65,6 +61,41 @@ export type CompletedChallenge = {
...
+    ({ id }) => challengeId === id
+  );
+
+  if (savedIndex >= 0) {
+    user.savedChallenges[savedIndex] = challengeToSave;
+  } else {
+    user.savedChallenges.push(challengeToSave);
```
Comment (ojeytonwilliams) [inline api/src/utils/common-challenge-functions.ts:93]:
```suggestion
    savedChallenges[savedIndex] = challengeToSave;
  } else {
    savedChallenges.push(challengeToSave);
```

File: api/src/utils/common-challenge-functions.ts [hunk 8]
```diff
@@ -65,6 +61,41 @@ export type CompletedChallenge = {
...
+  }
+
+  return user.savedChallenges;
```
Comment (ojeytonwilliams) [inline api/src/utils/common-challenge-functions.ts:96]:
```suggestion
  return savedChallenges;
```

File: api/src/utils/error-formatting.ts [hunk 9]
```diff
@@ -27,10 +27,12 @@ const getError = (errors: ErrorObject[]): ErrorObject => {
  * Format validation errors for /project-completed.
  *
  * @param errors An array of validation errors.
+ * @param message Unqiue message for the function.
```
Comment (ojeytonwilliams) [inline api/src/utils/error-formatting.ts:30]:
```suggestion
```
I'll tidy this up in another PR, but let's keep this function simple for now.

File: api/src/utils/error-formatting.ts [hunk 10]
```diff
@@ -27,10 +27,12 @@ const getError = (errors: ErrorObject[]): ErrorObject => {
...
  * @returns Formatted errors that can be used in the response.
  */
 export const formatProjectCompletedValidation = (
-  errors: ErrorObject[]
+  errors: ErrorObject[],
+  message?: FormattedError['message']
```
Comment (ojeytonwilliams) [inline api/src/utils/error-formatting.ts:35]:
```suggestion
  errors: ErrorObject[]
```

File: api/src/utils/error-formatting.ts [hunk 11]
```diff
@@ -43,7 +45,8 @@ export const formatProjectCompletedValidation = (
       }
     : {
         type: 'error',
-        message: 'That does not appear to be a valid challenge submission.'
+        message:
+          message ?? 'That does not appear to be a valid challenge submission.'
```
Comment (ojeytonwilliams) [inline api/src/utils/error-formatting.ts:49]:
```suggestion
        message: 'That does not appear to be a valid challenge submission.'
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: api/src/routes/challenge.ts [hunk 12]
```diff
@@ -367,5 +368,66 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
...
+    '/save-challenge',
+    {
+      schema: schemas.saveChallenge,
+      errorHandler(error, request, reply) {
+        if (error.validation) {
+          void reply.code(403);
+          return formatProjectCompletedValidation(
+            error.validation,
+            'That challenge type is not savable.'
+          );
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api/src/routes/challenge.ts]: Nope, that's a separate validation error. In the api-server `isValidChallengeCompletion` rejects requests if the id isn't an objectId or if `challengeType` exists and isn't a number.

File: api/src/routes/challenge.ts [hunk 13]
```diff
@@ -367,5 +368,63 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
...
+          id: challengeId,
+          files
+        };
+
+        if (!multifileCertProjectIds.includes(challengeId)) {
+          void reply.code(403);
+          return {
+            type: 'error',
+            message: 'That challenge type is not savable.'
+          } as const;
```
Comment (ojeytonwilliams) [inline api/src/routes/challenge.ts:400]:
```suggestion
          return 'That challenge type is not savable.';
```
It's silly and we will be changing this ASAP, but the api-server really does just return a string.

File: api/src/schemas.ts [hunk 14]
```diff
@@ -509,5 +504,28 @@ export const schemas = {
...
+          Type.Intersect([
+            saveChallengeBody,
+            Type.Object({ lastSavedDate: Type.Number() })
+          ])
+        )
+      }),
+      403: Type.Object({
+        type: Type.Literal('error'),
+        message: Type.Literal('That challenge type is not savable.')
+      }),
```
Comment (ojeytonwilliams) [inline api/src/schemas.ts:522]:
```suggestion
      403: Type.Literal('That challenge type is not savable.'),
```
See above.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Just one more thing!

**Review by ojeytonwilliams — APPROVED** [review]
Thanks, @Sboonny. LGTM :+1:

**Review by raisedadead — APPROVED** [review]

**Pull request merged**